src/lightning/pytorch/callbacks/model_checkpoint.py

```diff
@@ -742,6 +744,27 @@ def file_exists(self, filepath: _PATH, trainer: "pl.Trainer") -> bool:
         exists = self._fs.exists(filepath)
         return trainer.strategy.broadcast(exists)
 
+    def _should_remove_checkpoint(self, trainer: "pl.Trainer", previous: str, current: str) -> bool:
+        """Checks if the previous checkpoint should be deleted.
+
+        A checkpoint won't be deleted if it is a local file and any of the following cases apply:
+        - The previous checkpoint is the same as the current checkpoint
+        - The previous checkpoint is not in the current checkpoint directory
+        - The previous checkpoint is the checkpoint the Trainer resumed from
+
+        """
+        if previous == current:
+            return False
+        if self._fs.protocol != "file":
+            return True
+        previous = Path(previous).absolute()
+        resume_path = Path(trainer.ckpt_path).absolute() if trainer.ckpt_path is not None else None
+        if resume_path is not None and previous == resume_path:
+            return False
```

Comment:
This should be equivalent

```suggestion
        if trainer.ckpt_path is None:
            return False
        previous = Path(previous).absolute()
        if previous == Path(trainer.ckpt_path).absolute():
            return False
```

Reply:
Oh I was too quick, actually no it is not equivalent. With this code, we would return False too early under the condition `if trainer.ckpt_path is None`. We would still have to check the conditions below. The previous version was correct.